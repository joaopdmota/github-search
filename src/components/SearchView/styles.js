import styled from 'vue-styled-components';
import { Wrapper } from '../Layout';

export const Container = styled(Wrapper)`
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 40vh;
  background-color: #4158d0;
  background-image: linear-gradient(
    43deg,
    #4158d0 0%,
    #c850c0 46%,
    #ffcc70 100%
  );
`;
