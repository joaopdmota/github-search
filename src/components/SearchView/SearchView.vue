<template>
  <Wrapper direction="column">
    <Container justify="center" alignItems="center">
      <H1>Github profile search</H1>
      <Divider space="20" />
      <Wrapper>
        <b-form-input
          id="search_input"
          :state="state"
          v-model="search"
          placeholder="Enter your name"
          @keyup.enter="searchUser"
        ></b-form-input>
        <Separator space="10" />
        <b-button variant="dark" @click="searchUser" id="search_button">
          <b-spinner label="Spinning" v-if="loading"></b-spinner>
          <span v-else>
            Pesquisar
          </span>
        </b-button>
      </Wrapper>
    </Container>
    <Wrapper
      v-if="error"
      justify="center"
      direction="column"
      id="not_found_user"
    >
      <Divider space="20" />
      <b-alert show variant="danger">Falha ao pesquisar usuário</b-alert>
      <Divider space="20" />
    </Wrapper>
    <Wrapper
      id="user_info"
      v-if="user"
      alignItems="center"
      justify="center"
      direction="column"
    >
      <Divider space="20" />
      <Wrapper>
        <Img :src="user.avatar_url" />
        <Separator space="30" />
        <Wrapper direction="column" justify="flex-start">
          <Wrapper alignItems="center">
            <H1 color="black">Nome:</H1>
            <Separator space="5" />
            <H2 color="black">{{ user.name }}</H2>
          </Wrapper>
          <Wrapper alignItems="center">
            <H1 color="black">Usuário:</H1>
            <Separator space="5" />
            <H2 color="black">{{ user.login }}</H2>
          </Wrapper>
          <Wrapper alignItems="center">
            <H1 color="black">Seguidores:</H1>
            <Separator space="5" />
            <H2 color="black">{{ user.followers }}</H2>
          </Wrapper>
          <Wrapper alignItems="center">
            <H1 color="black">Repositórios:</H1>
            <Separator space="5" />
            <H2 color="black">{{ user.public_repos }}</H2>
          </Wrapper>

          <H1 color="black" size="16">Main repositories:</H1>
          <a
            v-for="(repo, index) in user.repos"
            :href="repo.html_url"
            :key="index"
          >
            {{ repo.name }}
          </a>
        </Wrapper>
      </Wrapper>
    </Wrapper>
  </Wrapper>
</template>

<script>
import { Container } from './styles';
import { H1, H2, Img } from '../Interface';
import { Wrapper, Divider, Separator } from '../Layout';
import { fetchRepositories, fetchUser } from '../../services';

export default {
  name: 'SearchView',
  data: () => ({
    search: '',
    state: null,
    user: null,
    loading: false,
    error: false,
  }),
  methods: {
    resetStates() {
      this.user = null;
      this.error = false;
    },
    validateUser() {
      if (this.user || this.error) this.resetStates();
      if (this.search !== '') {
        this.state = true;
        return this.state;
      }

      this.state = false;
      return this.state;
    },
    async searchUser() {
      if (this.validateUser()) {
        this.state = null;
        try {
          this.loading = true;
          const { data: user } = await fetchUser(this.search);
          this.user = user;
          const { data: repos } = await fetchRepositories(this.user.login);
          if (repos.length > 4) {
            repos.length = 4;
          }
          this.user.repos = repos;
        } catch {
          this.error = true;
        } finally {
          this.loading = false;
        }
      }
    },
  },
  components: {
    Container,
    H1,
    H2,
    Divider,
    Wrapper,
    Separator,
    Img,
  },
};
</script>

<style scoped>
button {
  box-shadow: unset !important;
}
</style>
