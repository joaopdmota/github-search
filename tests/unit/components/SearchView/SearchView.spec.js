import { mount } from '@vue/test-utils';
import nock from 'nock';
import SearchView from '../../../../src/components/SearchView/SearchView.vue';
import userFixture from './fixtures/user.json';
import reposFixture from './fixtures/repos.json';

describe('search view component', () => {
	test('renders the correct snapshot', () => {
		const wrapper = mount(SearchView);
		expect(wrapper).toMatchSnapshot();
	});

	describe('when user input some value', () => {
			describe('when github api returns an user', () => {
				beforeAll(() => {
					const expectedUser = 'joaopdmota';
					nock('https://api.github.com').get(`/users/${expectedUser}`).reply(200, { data: userFixture});
					nock('https://api.github.com').get(`/users/${expectedUser}/repos`).reply(200, { data: reposFixture});
				});

				test('should render user info', async () => {
					const wrapper = mount(SearchView);
					const input = wrapper.find('#search_input');
					await input.setValue('joaopdmota');
					const button = wrapper.find('#search_button');
					await button.trigger('click');
					expect(wrapper.vm.loading).toBe(true);

					wrapper.vm.$nextTick(() => {
						expect(wrapper.vm.loading).toBe(false);
						expect(wrapper.vm.error).toBe(false);
						expect(wrapper.find('#user_info').exists()).toBe(true);
						done();
					})
				});
			})

			describe('when github api returns an error', () => {
				// test('', () => {
				// 	const wrapper = mount(SearchView);
				// 	expect(wrapper).toMatchSnapshot();
				// });
			})
	})

	describe('when user does not input value', () => {
		// test('', () => {
		// 	const wrapper = mount(SearchView);
		// 	expect(wrapper).toMatchSnapshot();
		// });
	});
});