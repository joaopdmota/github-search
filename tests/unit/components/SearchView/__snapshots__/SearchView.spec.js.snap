// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search view component renders the correct snapshot 1`] = `
<div class="csogJw">
  <div class="hfazVC bHYHHw">
    <h1 class="eHwBcJ">Github profile search</h1>
    <div class="gDjLHi"></div>
    <div class="hxZbeR"><input id="search_input" type="text" placeholder="Enter your name" class="form-control">
      <div class="hWwYtV"></div> <button id="search_button" type="button" class="btn btn-dark"><span>
          Pesquisar
        </span></button>
    </div>
  </div>
  <!---->
  <!---->
</div>
`;
